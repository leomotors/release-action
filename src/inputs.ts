import * as core from "@actions/core";

import { Inputs, inputSchema } from "./schema/inputs.js";
import { TaskResult } from "./types.js";

export function getInput() {
  return {
    githubToken: core.getInput("github-token") || undefined,
    tag: core.getInput("tag") || undefined,
    mode: core.getInput("mode") || undefined,
    title: core.getInput("title") || undefined,
    changelogFile: core.getInput("changelog-file") || undefined,
    packagesInfoFile: core.getInput("packages-info-file") || undefined,
    dryRun: core.getBooleanInput("dry-run"),
  };
}

export function getInputAndParse(): TaskResult<Inputs> {
  try {
    const input = getInput();

    return {
      data: inputSchema.parse(input),
      success: true as const,
    };
  } catch (error) {
    return {
      error,
      success: false as const,
    };
  }
}
