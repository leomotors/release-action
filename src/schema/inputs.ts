import { z } from "zod";

export const tagRegex =
  /^(?:@?[\w\-/]+@)?v?\d+\.\d+\.\d+(?:-[a-zA-Z]+(?:\.\d+)?)?$/;

const tagSchema = z.string().regex(tagRegex);

const baseInputSchema = z.object({
  tag: tagSchema,
});

const releaseModeSchema = z
  .object({
    githubToken: z.string().nonempty(),
    title: z.string().nonempty().default("Release"),
    changelogFile: z.string().nonempty().default("CHANGELOG.md"),
    dryRun: z.boolean().default(false),
    mode: z.literal("release"),
  })
  .merge(baseInputSchema);

const getPackageInfoModeSchema = z
  .object({
    packagesInfoFile: z.string().nonempty().default("packages-info.yaml"),
    mode: z.literal("get-packages-info"),
  })
  .merge(baseInputSchema);

export const inputSchema = z.discriminatedUnion("mode", [
  releaseModeSchema,
  getPackageInfoModeSchema,
]);

export type Inputs = z.infer<typeof inputSchema>;
export type ReleaseModeInputs = z.infer<typeof releaseModeSchema>;
export type GetPackageInfoModeInputs = z.infer<typeof getPackageInfoModeSchema>;
