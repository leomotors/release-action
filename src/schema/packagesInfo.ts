import { z } from "zod";

const packageInfo = z.object({
  name: z.string(),
  // defaults to `name`
  fullName: z.string().optional(),
  packagePath: z.string().nonempty(),
  // defaults to `packagePath + "CHANGELOG.md"`
  changelogPath: z.string().nonempty().optional(),
});

export const packagesInfoSchema = z.object({
  packages: z.array(packageInfo).nonempty(),
});

export type PackageInfo = z.infer<typeof packageInfo>;
export type PackagesInfo = z.infer<typeof packagesInfoSchema>;
