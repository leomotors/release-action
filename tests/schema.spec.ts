import { describe, expect, it } from "vitest";
import type { ZodTypeAny } from "zod";

import { inputSchema, tagRegex } from "../src/schema/inputs.js";
import { packagesInfoSchema } from "../src/schema/packagesInfo.js";

const createExpectPass =
  (schema: ZodTypeAny) =>
  (input: unknown, checkSnapshot = true, checkEqual = false) => {
    expect(() => schema.parse(input)).not.toThrow();

    if (checkSnapshot) {
      expect(schema.parse(input)).toMatchSnapshot();
    }

    if (checkEqual) {
      expect(schema.parse(input)).toEqual(input);
    }
  };

const expectInputPass = createExpectPass(inputSchema);

const expectPkgPass = createExpectPass(packagesInfoSchema);

const createExpectFail =
  (schema: ZodTypeAny) =>
  (input: unknown, checkSnapshot = true) => {
    if (checkSnapshot) {
      expect(() => schema.parse(input)).toThrowErrorMatchingSnapshot();
    } else {
      expect(() => schema.parse(input)).toThrow();
    }
  };

const expectInputFail = createExpectFail(inputSchema);
const expectPkgFail = createExpectFail(packagesInfoSchema);

describe("Input Schema", () => {
  it("Valid Input: Release Mode", () => {
    // Simple Release
    expectInputPass({
      mode: "release",
      githubToken: "ghp_123",
      tag: "v1.0.0",
    });

    // Release with Title
    expectInputPass({
      mode: "release",
      githubToken: "ghp_123",
      tag: "v1.0.0",
      title: "Release Title",
    });

    // Release with Changelog
    expectInputPass({
      mode: "release",
      githubToken: "ghp_123",
      tag: "v1.0.0",
      changelogFile: "CHANGELOG-real.md",
    });

    // Dry Run
    expectInputPass({
      mode: "release",
      githubToken: "ghp_123",
      tag: "v1.0.0",
      dryRun: true,
    });
  });

  it("Valid Input: Get Packages Info Mode", () => {
    // Simple Get Packages Info
    expectInputPass({
      mode: "get-packages-info",
      githubToken: "ghp_123",
      tag: "v1.0.0",
    });

    // Override
    expectInputPass({
      mode: "get-packages-info",
      githubToken: "ghp_123",
      tag: "v1.0.0",
      packagesInfoFile: "packages-info.lmao",
    });
  });

  it("Invalid Input: No mode", () => {
    // Empty
    expectInputFail({});

    // Random Field
    expectInputFail({
      ruby: "chocomint",
      ayumu: "strawberry flavor",
      shiki: "cookie & cream",
    });

    // No Mode
    expectInputFail({
      githubToken: "ghp_123",
      tag: "v1.0.0",
    });

    // Invalid Mode
    expectInputFail({
      mode: "gx",
      githubToken: "ghp_123",
      tag: "v1.0.0",
    });
  });

  it("Invalid Input: Release Mode", () => {
    // No Tag
    expectInputFail({
      mode: "release",
      githubToken: "ghp_123",
    });

    // Empty String
    expectInputFail({
      mode: "release",
      githubToken: "ghp_123",
      tag: "",
    });
  });

  it("Invalid Input: Get Packages Info Mode", () => {
    // No Tag
    expectInputFail({
      mode: "get-packages-info",
      githubToken: "ghp_123",
    });
  });
});

describe("Tag Schema", () => {
  it("Valid Tag", () => {
    const tags = [
      "v1.0.0",
      "v1.2.3",
      "1.2.3",
      "1.2.3-rc",
      "1.2.3-rc.1",
      "4.5.6-beta.1",
      "3.0.4-alpha.1",
      "5.0.0",
      "0.0.0",
      "@package/nest/anothernest@6.9.0",
      "@package/nest/anothernest@6.9.0-rc.1",
      "@package/nest@6.9.0-beta.1",
      "package@6.9.0-alpha.1",
    ];

    tags.forEach((tag) => {
      expect(tagRegex.test(tag)).toBeTruthy();
    });
  });

  it("Invalid Tag", () => {
    const tags = [
      "1.2.3-rc.",
      "1.2.3-rc..1",
      "1.2.3-rc..",
      "1.2.3-rc..1.",
      "v1.2.3-rc.",
      "v1.2.3-rc..1",
      "v1.2.3-rc..",
      "v1.2.3-rc..1.",
      "@@v1.2.3",
      "@v1.2.3",
      "1.2",
      "1",
      "",
      "@",
    ];

    tags.forEach((tag) => {
      expect(tagRegex.test(tag)).toBeFalsy();
    });
  });
});

describe("Packages Info Schema", () => {
  it("Valid Case", () => {
    const ep = (input: unknown) => expectPkgPass(input, false, true);

    ep({
      packages: [
        {
          name: "hello",
          packagePath: "apps/hello",
        },
      ],
    });

    ep({
      packages: [
        {
          name: "hello",
          packagePath: "apps/hello",
          changelogPath: "apps/hello/not-changelog.md",
        },
      ],
    });

    ep({
      packages: [
        {
          name: "",
          packagePath: "apps/default",
        },
      ],
    });

    ep({
      packages: [
        {
          name: "hello",
          packagePath: "apps/hello",
        },
        {
          name: "world",
          fullName: "ワールド",
          packagePath: "apps/world",
          changelogPath: "apps/world/not-changelog.md",
        },
      ],
    });
  });

  it("Invalid Case", () => {
    expectPkgFail({
      name: "hello",
      packagePath: "apps/hello",
    });

    expectPkgFail({
      packages: [],
    });

    expectPkgFail({
      packages: [
        {
          name: "valid",
          packagePath: "apps/valid",
        },
        {
          name: "invalid",
          packagePath: "",
        },
      ],
    });

    expectPkgFail({
      packages: [
        {
          name: "valid",
          packagePath: "apps/valid",
        },
        {
          name: "invalid",
        },
      ],
    });

    expectPkgFail({
      packages: [
        {
          name: "valid",
          packagePath: "apps/valid",
        },
        {
          name: "invalid",
          packagePath: "apps/invalid",
          changelogPath: "",
        },
      ],
    });
  });
});
