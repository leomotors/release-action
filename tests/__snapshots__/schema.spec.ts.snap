// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input Schema > Invalid Input: Get Packages Info Mode 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "tag"
    ],
    "message": "Required"
  }
]]
`;

exports[`Input Schema > Invalid Input: Get Packages Info Mode 2`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "tag"
    ],
    "message": "Required"
  }
]]
`;

exports[`Input Schema > Invalid Input: No mode 1`] = `
[ZodError: [
  {
    "code": "invalid_union_discriminator",
    "options": [
      "release",
      "get-packages-info"
    ],
    "path": [
      "mode"
    ],
    "message": "Invalid discriminator value. Expected 'release' | 'get-packages-info'"
  }
]]
`;

exports[`Input Schema > Invalid Input: No mode 2`] = `
[ZodError: [
  {
    "code": "invalid_union_discriminator",
    "options": [
      "release",
      "get-packages-info"
    ],
    "path": [
      "mode"
    ],
    "message": "Invalid discriminator value. Expected 'release' | 'get-packages-info'"
  }
]]
`;

exports[`Input Schema > Invalid Input: No mode 3`] = `
[ZodError: [
  {
    "code": "invalid_union_discriminator",
    "options": [
      "release",
      "get-packages-info"
    ],
    "path": [
      "mode"
    ],
    "message": "Invalid discriminator value. Expected 'release' | 'get-packages-info'"
  }
]]
`;

exports[`Input Schema > Invalid Input: No mode 4`] = `
[ZodError: [
  {
    "code": "invalid_union_discriminator",
    "options": [
      "release",
      "get-packages-info"
    ],
    "path": [
      "mode"
    ],
    "message": "Invalid discriminator value. Expected 'release' | 'get-packages-info'"
  }
]]
`;

exports[`Input Schema > Invalid Input: No mode 5`] = `
[ZodError: [
  {
    "code": "invalid_union_discriminator",
    "options": [
      "release",
      "get-packages-info"
    ],
    "path": [
      "mode"
    ],
    "message": "Invalid discriminator value. Expected 'release' | 'get-packages-info'"
  }
]]
`;

exports[`Input Schema > Invalid Input: Release Mode 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "tag"
    ],
    "message": "Required"
  }
]]
`;

exports[`Input Schema > Invalid Input: Release Mode 2`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "githubToken"
    ],
    "message": "Required"
  }
]]
`;

exports[`Input Schema > Invalid Input: Release Mode 3`] = `
[ZodError: [
  {
    "validation": "regex",
    "code": "invalid_string",
    "message": "Invalid",
    "path": [
      "tag"
    ]
  }
]]
`;

exports[`Input Schema > Valid Input: Get Packages Info Mode 1`] = `
{
  "mode": "get-packages-info",
  "packagesInfoFile": "packages-info.yaml",
  "tag": "v1.0.0",
}
`;

exports[`Input Schema > Valid Input: Get Packages Info Mode 2`] = `
{
  "mode": "get-packages-info",
  "packagesInfoFile": "packages-info.lmao",
  "tag": "v1.0.0",
}
`;

exports[`Input Schema > Valid Input: Release Mode 1`] = `
{
  "changelogFile": "CHANGELOG.md",
  "dryRun": false,
  "githubToken": "ghp_123",
  "mode": "release",
  "tag": "v1.0.0",
  "title": "Release",
}
`;

exports[`Input Schema > Valid Input: Release Mode 2`] = `
{
  "changelogFile": "CHANGELOG.md",
  "dryRun": false,
  "githubToken": "ghp_123",
  "mode": "release",
  "tag": "v1.0.0",
  "title": "Release Title",
}
`;

exports[`Input Schema > Valid Input: Release Mode 3`] = `
{
  "changelogFile": "CHANGELOG-real.md",
  "dryRun": false,
  "githubToken": "ghp_123",
  "mode": "release",
  "tag": "v1.0.0",
  "title": "Release",
}
`;

exports[`Input Schema > Valid Input: Release Mode 4`] = `
{
  "changelogFile": "CHANGELOG.md",
  "dryRun": true,
  "githubToken": "ghp_123",
  "mode": "release",
  "tag": "v1.0.0",
  "title": "Release",
}
`;

exports[`Packages Info Schema > Invalid Case 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "packages"
    ],
    "message": "Required"
  }
]]
`;

exports[`Packages Info Schema > Invalid Case 2`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "array",
    "inclusive": true,
    "exact": false,
    "message": "Array must contain at least 1 element(s)",
    "path": [
      "packages"
    ]
  }
]]
`;

exports[`Packages Info Schema > Invalid Case 3`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "packages",
      1,
      "packagePath"
    ]
  }
]]
`;

exports[`Packages Info Schema > Invalid Case 4`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "packages",
      1,
      "packagePath"
    ],
    "message": "Required"
  }
]]
`;

exports[`Packages Info Schema > Invalid Case 5`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "packages",
      1,
      "changelogPath"
    ]
  }
]]
`;
